src/main/java/ca/corefacility/bioinformatics/irida/database/changesets/FastqcToFilesystem.java

```diff
@@ -0,0 +1,248 @@
+
+/**
+ * Liquibase update class for moving the fastqc analysis results off the {@link ca.corefacility.bioinformatics.irida.model.workflow.analysis.AnalysisFastQC}
+ * class (and out of the database).  This will instead save them to the filesystem as an {@link
+ * ca.corefacility.bioinformatics.irida.model.workflow.analysis.AnalysisOutputFile}.  This should greatly decrease the
+ * size of the database for large IRIDA installs and speed up loading of the {@link
+ * ca.corefacility.bioinformatics.irida.model.workflow.analysis.AnalysisFastQC} class.
+ */
+public class FastqcToFilesystem implements CustomSqlChange {
+
+	private static final Logger logger = LoggerFactory.getLogger(FastqcToFilesystem.class);
+	private Path outputFileDirectory;
+	private Path tempOutputDirectory;
+
+	private DataSource dataSource;
+
+	@Override
+	public SqlStatement[] generateStatements(Database database) throws CustomChangeException {
+
+		JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
+
+		List<Path> writtenFiles = new ArrayList<>();
+
+		//create a temp directory for fastqc files
+		try {
+			this.tempOutputDirectory = Files.createTempDirectory(outputFileDirectory, "irida-fastqc-temp");
+		} catch (IOException e) {
+			throw new CustomChangeException("Could not create temp directory for fastqc files", e);
+		}
+
+		logger.info("This update will be writing files temporarily to " + tempOutputDirectory
+				+ ".  They should be automatically transferred to your analysis output directory " + outputFileDirectory
+				+ "  upon completion.");
+
+		//wrapping everything in a try/catch so we can log a nice error message in case of a problem.  Err
+		try {
+
+			//the chart types we're taking out of the database
+			List<String> chartTypes = Lists.newArrayList("perBaseQualityScoreChart", "perSequenceQualityScoreChart",
+					"duplicationLevelChart");
+
+			//inserting empty output files for charts
+			chartTypes.forEach(chart -> {
+				int update = jdbcTemplate.update(
+						"INSERT INTO analysis_output_file (created_date, execution_manager_file_id, analysis_id) SELECT createdDate, '"
+								+ chart + ".png', a.id FROM analysis a INNER JOIN analysis_fastqc q ON a.id=q.id");
+
+				logger.info("Inserted " + update + " temp output file entries for chart type " + chart);
+			});
+
+			//for each chart type, write the new files
+			chartTypes.forEach(chart -> {
+				//write the files and update the analysis_output_file directory
+				writtenFiles.addAll(writeFileForChartType(chart, jdbcTemplate));
+
+				//Add entries to analysis_output_file_map linking to the new analysis_output_file entries
+				jdbcTemplate.update(
+						"insert into analysis_output_file_map (analysis_id, analysisOutputFilesMap_id, analysis_output_file_key) SELECT analysis_id, id, '"
+								+ chart + "' FROM analysis_output_file where execution_manager_file_id='" + chart
+								+ ".png'");
+			});
+
+			//if everything went well, we can move those files into the real output directory
```

Comment:
In theory, the code starting from here should be outside of this try/catch block.

The reason being, if there's an exception in the middle of moving files from the temp directory, the statement `You do *not* have to cleanup existing FastQC images on the filesystem` no longer applies (since you will have to clean up files on the file system in this case).

So, maybe move this block of code outside the main try/catch. If there's an error moving files, log an error message containing the starting analysis output file id (e.g., `12345`). Say that users must delete all folders in the analysis output directory starting from `12345`, restore the database from backup, and try again.

Reply:
Good idea.  This is done in 3735258